src/Build/BackEnd/Components/RequestBuilder/RequestBuilder.cs

```diff
@@ -1174,7 +1174,7 @@ private ProjectInstance LoadProjectFromFile()
                 _componentHost.BuildParameters,
                 _nodeLoggingContext.LoggingService,
                 new BuildEventContext(
-                    _requestEntry.Request.BuildEventContext.SubmissionId,
+                    _requestEntry.Request.SubmissionId,
```

Comment:
Is it correct for `_requestEntry.Request.BuildEventContext.SubmissionId` to be invalid but `_requestEntry.Request.SubmissionId` valid? If both should be valid, then maybe it's worth fixing that.

Reply:
I've thought about that. The trouble is that BuildEventContext gets reset a few times on the way there, for instance here: https://source.dot.net/#Microsoft.Build/BackEnd/Components/RequestBuilder/RequestBuilder.cs,1094
or here:
https://source.dot.net/#Microsoft.Build/BackEnd/Shared/BuildRequest.cs,132

Reply:
In fact this constructor is a direct counterexample to your hypothesis that they should be in sync: it has a submission ID but invalid build event context:
https://source.dot.net/#Microsoft.Build/BackEnd/Shared/BuildRequest.cs,132

Reply:
Hmm, from your link, it seems that the RequestBuilder is the one that gives the BuildRequest a valid BuildEventContext: https://source.dot.net/#Microsoft.Build/BackEnd/Components/RequestBuilder/RequestBuilder.cs,1094. So during a build, there's a valid submission ID, but not during the evaluation. The evaluation events don't get the submission ID because the submission ID along with the valid BuildEventContext get generated at line 1081 whereas the evaluation happens before that, at line 1063.

So I'd say your fix is fine, with the observation that if future evaluation events would access other BuildEventContext IDs generated at line 1081 in the RequestBuilder, then this fix will have to be changed again.

Reply:
Alternatively, the fix to fix all bugs of this type would be to separate the generation of submission ID from the logging of ProjectStarted events, so that the evaluation ca use the valid BuildEventContext.

Reply:
You're right, it's a bit of a chicken and egg. You need the BuildEventContext from ProjectLoggingContext, but that is created later.